docs/design/security_context.md

```diff
@@ -0,0 +1,158 @@
+  [service accounts](https://github.com/GoogleCloudPlatform/kubernetes/pull/2297).
+
+## Use Cases
+
+In order of increasing complexity, following are example use cases that would 
+be addressed with security contexts:
+
+1.  Kubernetes is used to run a single cloud application. In order to protect
+    nodes from containers:
+    * All containers run as a single non-root user
+    * Privileged containers are disabled
+    * All containers run with a particular MCS label 
+    * Kernel capabilities like CHOWN and MKNOD are removed from containers
+    
+2.  Just like case #1, except that I have more than one application running on
+    the Kubernetes cluster.
+    * Each application is run in its own namespace to avoid name collisions
+    * For each application a different uid and MCS label is used
+    
+3.  Kubernetes is used as the base for a PAAS with 
+    multiple projects, each project represented by a namespace. 
+    * Each namespace is associated with a range of uids/gids on the node that
+      are mapped to uids/gids on containers using linux user namespaces. 
+    * Certain pods in each namespace have special privileges to perform system
+      actions such as talking back to the server for deployment, run docker
+      builds, etc.
+    * External NFS storage is assigned to each namespace and permissions set
+      using the range of uids/gids assigned to that namespace. 
+
+## Proposed Design
+
+### Overview
+A *security context* consists of a set of constraints that determine how a container
+is secured before getting created and run. It has a 1:1 correspondence to a
+[service account](https://github.com/GoogleCloudPlatform/kubernetes/pull/2297). A *security context provider* is passed to the Kubelet so it can have a chance
+to mutate Docker API calls in order to apply the security context.
+
+It is recommended that this design be implemented in two phases:
+
+1.  Implement the security context provider extension point in the Kubelet 
+    so that a default security context can be applied on container run and creation.
+2.  Implement a security context structure that is part of a service account. The
+    default context provider can then be used to apply a security context based
+    on the service account associated with the pod.
+    
+### Security Context Provider
+
+The Kubelet will have an interface that points to a `SecurityContextProvider`. The `SecurityContextProvider` is invoked before creating and running a given container:
+
+```go
+type SecurityContextProvider interface {
+    // ModifyContainerConfig is called before the Docker createContainer call.
+    // The security context provider can make changes to the Config with which
+    // the container is created.
+    // An error is returned if it's not possible to secure the container as 
+    // requested with a security context. 
+	ModifyContainerConfig(pod *api.BoundPod, container *api.Container, config *docker.Config) error
```

Comment:
----- Original Message -----
> 
> 
> ----- Original Message -----
> > > +2.  Implement a security context structure that is part of a service
> > > account. The
> > > +    default context provider can then be used to apply a security
> > > context
> > > based
> > > +    on the service account associated with the pod.
> > > +
> > > +### Security Context Provider
> > > +
> > > +The Kubelet will have an interface that points to a
> > > `SecurityContextProvider`. The `SecurityContextProvider` is invoked
> > > before
> > > creating and running a given container:
> > > +
> > > +```go
> > > +type SecurityContextProvider interface {
> > > +    // ModifyContainerConfig is called before the Docker createContainer
> > > call.
> > > +    // The security context provider can make changes to the Config with
> > > which
> > > +    // the container is created.
> > > +    // An error is returned if it's not possible to secure the container
> > > as
> > > +    // requested with a security context.
> > > +	ModifyContainerConfig(pod *api.BoundPod, container *api.Container,
> > > config
> > > *docker.Config) error
> > 
> > Would it work to have the SecurityContextProvider just modify the
> > api.BoundPod, and not take a docker.Config as an argument.  Reasons:
> >  - follow the pattern already in the code where we modify objects as they
> >  are
> >  passed along, including how we add env vars to the pod object in the
> >  kubelet.
> >  - we probably will want a kubelet debug API that lets you see the "actual
> >  Pod started", including env vars and security context modifications
> >  - we want to be able to bootstrap using http or file sources.  So, this
> >  ensures that any security context information is expressible in the pod.
> >  - It makes it a tad easier to put in a docker alternative eventually if
> >  docker.Config is not used in as many places in the code.
> 
> True - I think cesar (correct me if I'm wrong here) had started here because
> the options to docker may be complex - setting up user namespaces, labels,
> and default behavior. However, a two step abstraction of making the docker
> interface we have from the kubelet support the additional options on
> BoundPods, and then adding bound pods options, seems reasonable.

One thing I did think of - you may need to know (from the image) what user the image is going to run as, and like ENTRYPOINT I think it's frustrating to an end user to have to specify that image up front in the pod definition.  Some level of "map user X inside the container to Y outside" happening by default seemed potentially valuable.  However, the two step process (setup security context, then pass to the docker interface) could also handle that.

> 
> Some security context stuff might be a finalizer at the master level.
> Security context, if applied on the kubelet for final defaults, and on the
> master for cluster level isolation, seems similar to other finalizer style
> patterns.